src/turbine/retransmit_service.zig

```diff
@@ -0,0 +1,432 @@
+const std = @import("std");
+const net = @import("zig-network");
+const sig = @import("../sig.zig");
+
+const socket_utils = sig.net.socket_utils;
+
+const Random = std.rand.Random;
+const UdpSocket = net.Socket;
+const EndPoint = net.EndPoint;
+const AtomicBool = std.atomic.Value(bool);
+const AtomicU64 = std.atomic.Value(u64);
+
+const Packet = sig.net.Packet;
+const Pubkey = sig.core.Pubkey;
+const Slot = sig.core.Slot;
+const ThreadSafeContactInfo = sig.gossip.data.ThreadSafeContactInfo;
+const Counter = sig.prometheus.Counter;
+const Gauge = sig.prometheus.Gauge;
+const GetMetricError = sig.prometheus.registry.GetMetricError;
+const Duration = sig.time.Duration;
+const TurbineTree = sig.turbine.TurbineTree;
+const TurbineTreeCache = sig.turbine.TurbineTreeCache;
+const Channel = sig.sync.Channel;
+const ShredId = sig.ledger.shred.ShredId;
+const LeaderScheduleCache = sig.core.leader_schedule.LeaderScheduleCache;
+const BankFields = sig.accounts_db.snapshots.BankFields;
+const RwMux = sig.sync.RwMux;
+const Logger = sig.trace.log.Logger;
+const ShredDeduper = sig.turbine.shred_deduper.ShredDeduper;
+
+const globalRegistry = sig.prometheus.globalRegistry;
+
+/// Shred deduper configuration from agave
+const DEDUPER_MAX_DUPLICATE_COUNT: usize = 2;
+const DEDUPER_FALSE_POSITIVE_RATE: f64 = 0.001;
+const DEDUPER_RESET_CYCLE: Duration = Duration.fromSecs(5 * 60);
+const DEDUPER_NUM_BITS: u64 = 637_534_199;
+
+/// Retransmit Service
+/// The retransmit service receives verified shreds from the shred collector and retransmits them to the network.
+/// The retransmit service is broken down into two main components:
+/// 1. receiveShreds: runs on a single thread and receives shreds from the shred collector, deduplicates them, and then packages them
+///    into RetransmitShredInfo's which are sent to a channel for further processing.
+/// 2. retransmitShreds: runs on N threads and receives RetransmitShredInfo's from the channel, computes the children to retransmit to
+///    and then constructs and sends packets to the network.
+pub fn run(
+    allocator: std.mem.Allocator,
+    my_contact_info: ThreadSafeContactInfo,
+    bank_fields: *const BankFields,
+    leader_schedule_cache: *LeaderScheduleCache,
+    gossip_table_rw: *RwMux(sig.gossip.GossipTable),
+    receiver: *Channel(Packet),
+    num_retransmit_sockets: usize,
+    maybe_num_retransmit_threads: ?usize,
+    overwrite_stake_for_testing: bool,
+    exit: *AtomicBool,
+    rand: Random,
+    logger: Logger,
+) !void {
```

Comment:
Are named parameter structs something we have decided to adhere to? I'm not sold on them as I think they add an additional layer of indirection, when looking at the function definition you then need to go and look at a parameter struct to know what it depends on. If it's an established approach we are taking then I am happy to conform but I don't love the pattern. To be clear I don't mind the use of options structs when there are a lot of bools and ints (for example the run method could take an options struct for the num_retransmit_sockets, num_retransmit_threads, and overwrite_stake_for_testing) but putting all the parameters into struct doesn't make sense to me.

Reply:
I don't think it's something we have to universally adhere to, rather I prefer to evaluate things on a case-by-case basis. I request this specifically when I believe it to be a benefit to the readability of the callsite, which I would argue is of much greater importance and value than reading the definition: 99% of the time when I'm reading code, I don't jump to every function's definition, I just try to get a general understanding from how they're called, unless really necessary. These long unnamed parameter lists make it so I basically *have* to read the function definition to understand what everything means, whereas a named parameter wouldn't leave me guessing what `true` means, because it would be `.overwrite_stake_for_testing = true`.

If your main issue with named parameter structs is the indirection, a compromise I would propose is inlining the struct definition, especially applicable when it's a parameter list like this which is unlikely to be directly forwarded by anything else:
```suggestion
pub fn run(params: struct {
    allocator: std.mem.Allocator,
    my_contact_info: ThreadSafeContactInfo,
    bank_fields: *const BankFields,
    leader_schedule_cache: *LeaderScheduleCache,
    gossip_table_rw: *RwMux(sig.gossip.GossipTable),
    receiver: *Channel(Packet),
    num_retransmit_sockets: usize,
    maybe_num_retransmit_threads: ?usize,
    overwrite_stake_for_testing: bool,
    exit: *AtomicBool,
    rand: Random,
    logger: Logger,
}) !void {
```

(P.S. sorry for the wall of text)

Reply:
9982c1f5